### PR TITLE
Inference Optimized MoEs

### PR DESCRIPTION
**Pull request by sidsingh-nvidia**
# What does this PR do ?
Optimizes MoE decode (and small batch prefill) performance by eliminating host synchronizations, enabling end-to-end CUDA graph capture, and using latency-optimized NVLS collectives.

### Motivation

The default MoE layer in Megatron-LM is not well-suited for inference. The AlltoAll dispatcher and GroupedGEMM rely on CPU-resident tensors for token-expert assignments, which breaks CUDA graph capture due to host synchronizations. The current workaround — padding to maximum capacity (routing all tokens to all experts) — enables static shapes but wastes significant compute and communication.

This PR introduces an inference-optimized MoE layer that achieves the best of both worlds: compute/communication-optimal routing with full CUDA graph compatibility. These layers are specially designed to optimize the decode phase, where small batch sizes make host synchronization and kernel launch overhead disproportionately expensive.

### Optimizations

- **`InferenceGroupedMLP`** — Expert computation layer that operates directly on GPU-resident token-expert splits, eliminating the host synchronizations that block CUDA graph capture in the default GroupedGEMM path. Uses FlashInfer `cutlass_fused_moe` (fused permute + GEMM) for CUDA-graphed iterations and `torch._grouped_mm` with GPU-resident cumsum offsets for eager mode. Inherits from `TEGroupedMLP` for checkpoint compatibility.. 
- **`InferenceTopKRouter`** — Stripped-down router that removes training overhead (z-loss, auxiliary losses, token dropping) and is optimized via `@torch.compile()`.
- **`InferenceCUDAGraphTokenDispatcher`** — Replaces AlltoAll with AllGather/ReduceScatter for token exchange, keeping all metadata GPU-resident. Supports latency-optimized NVLS collectives on Hopper+ with automatic NCCL fallback.
- **Fused 3-tensor NVLS all-gather kernel** (`multimem_all_gather_fused`) for routing_map, probs, and hidden_states — single kernel launch + single barrier. 

### Other Minor Changes

- **`InferenceSpecProvider`** backend for wiring inference-optimized modules into model specs.
- **`MoELayer`** dynamically swaps between standard and inference dispatchers based on `is_inference_cuda_graphed_iteration`.
- Centralized NVLS eligibility checks (`are_tensors_nvls_eligible`, `is_device_nvls_capable`) shared across TP and EP communication paths.
- Separate symmetric memory buffer pools for TP and EP (`get_global_symmetric_memory_buffer_tp/ep`).
- Multi-tensor packing in symmetric memory buffers (`maybe_get_tensors`) with 16-byte alignment.
- Kill-switch via `--inference-disable-triton-nvls-kernels` config flag. This makes the system fallback to NCCL.
- Config validation: inference-optimized MoE rejects expert tensor parallelism, capacity-factor routing, and padded routing maps.

### How to  enable? 

these flags - 
```
--transformer-impl inference_optimized \
--moe-router-dtype fp32 # flashinfer only supports fp32 probabilities
```

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/transformer/moe/experts.py [hunk 1]
```diff
@@ -912,6 +920,230 @@ def backward_dw(self):
...
+        assert HAVE_FLASHINFER, "flashinfer-python is required for FlashInfer forward path."
+        output = fused_moe.cutlass_fused_moe(
+            hidden_states,
+            routing_map.to(torch.int),
+            probs.float(),
+            self._fc1_weight,
+            self._fc2_weight,
+            hidden_states.dtype,
+            quant_scales=None,
+            activation_type=ActivationType.Relu2,
```
Comment (kvareddy) [inline megatron/core/transformer/moe/experts.py]: What other activation types does cutlass_fused_moe support?

File: megatron/core/transformer/moe/moe_layer.py [hunk 2]
```diff
@@ -246,10 +263,63 @@ def __init__(
...
+
+    def _setup_inference_mode(self, pg_collection):
+        """Set up inference-optimized token dispatcher and state.
+
+        Called from __init__ when config.transformer_impl == "inference_optimized".
+        Creates an InferenceAllGatherTokenDispatcher alongside the standard dispatcher,
+        which is swapped in during CUDA-graphed forward passes.
+        """
+        
+        assert self.config.moe_token_dispatcher_type == "alltoall", (
```
Comment (kvareddy) [inline megatron/core/transformer/moe/moe_layer.py:294]: shouldn't this be allgather dispatcher here?

File: megatron/core/transformer/moe/experts.py (hunk 1 above)
Comment (sidsingh-nvidia) [reply megatron/core/transformer/moe/experts.py]: Yeah good point, they support swiglu silu etc. I should make this more general.

File: megatron/core/transformer/moe/moe_layer.py (hunk 2 above)
Comment (sidsingh-nvidia) [reply megatron/core/transformer/moe/moe_layer.py:294]: For training, I want to keep using  the all to all dispatcher. I am treating `--moe-token-dispatcher-type` as the training side argument for backwards compatibility,

**Comment by jaredcasper** [comment]
LGTM, will leave final ADLR approval to Vijay.

File: megatron/core/transformer/moe/experts.py [hunk 3]
```diff
@@ -720,6 +729,63 @@ def _apply_bias(intermediate_parallel, bias_parallel, tokens_per_expert, permute
...
+                    bias_parallel,
+                    permuted_probs,
+                    self.config.activation_func_fp8_input_store,
+                    self.config.glu_linear_offset,
+                    self.config.activation_func_clamp_value,
+                )
+            else:
+                raise ValueError("Only support fusion of swiglu and quick_gelu in TEGroupedMLP.")
+        elif self.activation_func == squared_relu and self.config.use_fused_weighted_squared_relu:
+            assert bias_parallel is None
```
Comment (Wohox) [inline megatron/core/transformer/moe/experts.py]: In the original code, error message is skipped, please add assertion error message with your refactoring, thanks.

File: megatron/core/transformer/moe/experts.py [hunk 4]
```diff
@@ -912,6 +919,202 @@ def backward_dw(self):
...
+        permuted_local_hidden_states: torch.Tensor,
+        tokens_per_expert: torch.Tensor,
+        permuted_probs: torch.Tensor,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
+        """Forward pass with three modes:
+        - Training: delegates to parent TEGroupedMLP
+        - Inference + CUDA graphed: FlashInfer cutlass_fused_moe
+        - Inference + eager: torch._grouped_mm with GPU-resident offsets
+        """
+        if self.training:
```
Comment (Wohox) [inline megatron/core/transformer/moe/experts.py:1111]: Wondering why is training a fallback option, shouldn't it be `assert not self.training`?

File: megatron/core/transformer/moe/experts.py [hunk 5]
```diff
@@ -912,6 +919,202 @@ def backward_dw(self):
...
+            ep_size=self.ep_group.size(),
+            ep_rank=self.ep_group.rank(),
+        )[0]
+        return output, None
+
+    def _torch_grouped_mm_forward(
+        self, permuted_local_hidden_states, tokens_per_expert, permuted_probs
+    ):
+        permuted_probs = permuted_probs.unsqueeze(-1)
+        # assert tokens_per_expert.is_cuda, "tokens_per_expert must be on GPU"
```
Comment (Wohox) [inline megatron/core/transformer/moe/experts.py]: Please remove useless comments.

**Review by kvareddy — APPROVED** [review]

**Review by kvareddy — APPROVED** [review]

File: megatron/core/transformer/moe/moe_layer.py [hunk 6]
```diff
@@ -246,10 +266,76 @@ def __init__(
...
 
+    def _setup_inference_mode(self, pg_collection):
+        """Set up inference-optimized token dispatcher and state.
+
+        Called from __init__ when config.transformer_impl == "inference_optimized".
+        Creates an InferenceCUDAGraphTokenDispatcher alongside the standard dispatcher,
+        which is swapped in during CUDA-graphed forward passes.
+        """
+
+        assert self.config.moe_token_dispatcher_type == "alltoall", (
```
Comment (Wohox) [inline megatron/core/transformer/moe/moe_layer.py:294]: In training phase, why aren't `flex` and `all gather` dispatcher supported?

**Review by Wohox — APPROVED** [review]
Approved in advance, left some minor comments.

File: megatron/core/transformer/moe/router.py [hunk 7]
```diff
@@ -710,3 +710,91 @@ def _save_to_state_dict(self, *args, **kwargs):
...
+        )
+        assert config.moe_router_score_function in ["sigmoid", "softmax"], (
+            f"InferenceTopKRouter requires moe_router_score_function in "
+            f"['sigmoid', 'softmax'], got '{config.moe_router_score_function}'"
+        )
+
+        super().__init__(config=config, pg_collection=pg_collection)
+
+        self.is_inference_cuda_graphed_iteration = False
+        self.topk_routing_with_score_function = torch.compile(topk_routing_with_score_function)
```
Comment (Wohox) [inline megatron/core/transformer/moe/router.py]: Minor suggestion, use decorator instead, otherwise better to add assertion check for `TORCHINDUCTOR_FORCE_DISABLE_CACHES=0`.

File: megatron/core/inference/contexts/dynamic_context.py [hunk 8]
```diff
@@ -538,15 +538,18 @@ def __init__(self, model_config: TransformerConfig, inference_config: InferenceC
...
+                cuda_graph_mixed_prefill_request_count=inference_config.cuda_graph_mixed_prefill_count,
                 max_requests=self.max_requests,
                 max_tokens=self.max_tokens,
                 max_sequence_length=self.max_sequence_length,
                 use_cuda_graphs_for_non_decode_steps=self.use_cuda_graphs_for_non_decode_steps,
             )
         )
 
-        self.cuda_graph_mixed_prefill_count = inference_config.cuda_graph_mixed_prefill_count
+        self.smallest_non_decode_cuda_graph_size = (
```
Comment (buptzyb) [inline megatron/core/inference/contexts/dynamic_context.py]: Why is this a tuple? I think it should be an int?

File: megatron/core/transformer/moe/token_dispatcher_inference.py [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+            probs = gather_from_sequence_parallel_region(probs, group=self.tp_ep_group)
+            hidden_states = gather_from_sequence_parallel_region(
+                hidden_states, group=self.tp_ep_group
+            )
+
+        return hidden_states, probs
+
+    def dispatch_postprocess(self, hidden_states, probs):
+        """Pass-through: returns unpermuted inputs and routing_map for InferenceGroupedMLP."""
+        return hidden_states, self.routing_map, probs
```
Comment (buptzyb) [inline megatron/core/transformer/moe/token_dispatcher_inference.py]: The second returned value of `dispatch_postprocess` should be`tokens_per_expert`. If we need `self.routing_map` to be returned, I'd suggest modifying the return value signature of `dispatch_postprocess`: `return hidden_states, tokens_per_expert (None here), probs, self.routing_map`. It will be very confusing for future readers if we return the second value as different things in different cases.

File: megatron/core/transformer/moe/experts.py (hunk 3 above)
Comment (sidsingh-nvidia) [reply megatron/core/transformer/moe/experts.py]: addressed

File: megatron/core/transformer/moe/experts.py (hunk 4 above)
Comment (sidsingh-nvidia) [reply megatron/core/transformer/moe/experts.py:1111]: We want these layers for colocated RL and thus need them to toggle between inference and training mode.

File: megatron/core/transformer/moe/experts.py (hunk 5 above)
Comment (sidsingh-nvidia) [reply megatron/core/transformer/moe/experts.py]: done!

File: megatron/core/transformer/moe/moe_layer.py (hunk 6 above)
Comment (sidsingh-nvidia) [reply megatron/core/transformer/moe/moe_layer.py:294]: We haven't tested flex and all-gather (for training) in megatron-rl and nemo-rl (with mcore-inference) yet. We plan to enable these after testing.

File: megatron/core/transformer/moe/router.py (hunk 7 above)
Comment (sidsingh-nvidia) [reply megatron/core/transformer/moe/router.py]: done

File: megatron/core/inference/contexts/dynamic_context.py (hunk 8 above)
Comment (sidsingh-nvidia) [reply megatron/core/inference/contexts/dynamic_context.py]: Thank you so much for catching this Robin! This was a sneaky bug :)

File: megatron/core/transformer/moe/token_dispatcher_inference.py (hunk 9 above)
Comment (sidsingh-nvidia) [reply megatron/core/transformer/moe/token_dispatcher_inference.py]: fixed!

**Review by jaredcasper — APPROVED** [review]
